feat: signup tokens

SHAcollision opened the pull request:
Fixes #74

- [x] admin password and new config
- [x] new table for signup_tokens
- [x] new `/admin/generate_signup_token` endpoint
- [x] homeserver can run with and without signup token requirements
- [x] client can signup with a valid token. Fails to signup with an invalid token. And can signup always when no token required.
- [x] signup is not a wrapper of signin anymore. Now, sign in does not create new users (lmao). Sign up will return an error if user existed already.
- [x] Client test for native signup flow with a signup token.
- [x] Wasm API updated and wasm building
- [x] Client test for wasm signup flow with a signup token.

SHAcollision commented:
@SeverinAlexB made a big push here. All the basic functionality is in place. Lots of room to improve and clean up here and there. Consider it a draft to iterate over, but it would already benefit from a review and another pair of eyes on it.

Managed to build the wasm and API is alright, it's untested though. Will try to implement wasm tests for signup with tokens tmr.

SHAcollision commented on pubky-homeserver/src/core/database/tables/entries.rs:
```diff
@@ -431,7 +431,7 @@ impl<'db> EntryWriter<'db> {
 }
 
-impl<'db> std::io::Write for EntryWriter<'db> {
+impl std::io::Write for EntryWriter<'_> {
```
`clippy` did these. I guess it makes sense?

SeverinAlexB replied:
I think that's ok but not sure.

SeverinAlexB commented:
Something more that comes to my mind: What are John requirements again? Do tokens expire? Do we need the ability to deactivate/delete them?

SHAcollision commented on pubky-homeserver/src/config.example.toml:
```diff
@@ -1,6 +1,14 @@
+[admin]
+# Set an admin password to protect admin endpoints.
+# If no password is set, the admin endpoints will not be accessible. 
+password = "mySecretAdminPassword"
```
I agree. I take this is a very early iteration of "admin passwords" with the only purpose of getting signup codes working. A proper nominative admin system will be needed eventually to know "who performed this admin action" as this seems to be a requirement.

This file needs to be kept secret in its current design because it also has the homeserver private key in it. That's arguably even more sensitive than admin password because once leaked, it cannot be changed.

Let me know your thoughts, because my feeling is to leave it "as is" but we can of course be improve this. Storing a plaintext password in a config file is not that unusual and potential users likely know how to keep this file out of public access, after all, Bitcoind, `bitcoin.conf`, MySql `my.cnf`, rabbitmq.conf, reddis.conf, Django's settings.py, and many other store unsalted plaintext passwords in one way or another.

SeverinAlexB replied:
Let's keep it simple in this PR. We do have some time for refactors and we can check more advanced secret strategies then.

SHAcollision commented on pubky-homeserver/src/config.example.toml:
```diff
@@ -1,6 +1,14 @@
+# Set an admin password to protect admin endpoints.
+# If no password is set, the admin endpoints will not be accessible. 
+password = "mySecretAdminPassword"
+# Set signup_mode to "open" to allow anyone to signup a new user,
+# otherwise, "closed" (the default) to require a valid invite token on signup;
+signup_mode = "closed"
```
What do you think about `restricted` ? I think it's my OCD but if it's only 2 options I like to keep it Enums to 1 word. We could make it a `bool` too. `signup_open = false`

SHAcollision commented on pubky-homeserver/src/config.rs:
```diff
@@ -175,14 +187,26 @@ impl TryFrom<ConfigToml> for Config {
             }
         };
 
+        let admin = if let Some(admin_toml) = value.admin {
+            AdminConfig {
+                password: admin_toml.password.clone(),
+                signup_mode: match admin_toml.signup_mode.as_deref() {
+                    Some("open") => SignupMode::Open,
+                    _ => SignupMode::Closed,
+                },
+            }
+        } else {
+            AdminConfig::default()
+        };
```
Yes, agree. I thought about this as well. It's clucky that each config struct needs 2 variants `*Config` and `*ConfigToml` . Though I get this is because not all config is intended to be set from the `.toml` file.

SHAcollision commented on pubky-homeserver/src/core/database/tables/signup_tokens.rs:
```diff
@@ -0,0 +1,87 @@
+
+impl DB {
+    pub fn generate_signup_token(&mut self) -> anyhow::Result<String> {
+        let token = generate_random_token();
+        let signup_token = SignupToken {
+            token: token.clone(),
+            created_at: Timestamp::now().as_u64(),
+            used: None,
+        };
```
Nice!

SHAcollision commented on pubky-homeserver/src/core/routes/admin.rs:
```diff
@@ -0,0 +1,48 @@
+    response::IntoResponse,
+};
+
+pub async fn generate_signup_token(
+    State(mut state): State<AppState>,
+    headers: HeaderMap,
+) -> Result<impl IntoResponse> {
```
Yes, we can call this endpoint from anywhere we want. I will document how to generate a new token using `curl` .

SHAcollision commented on pubky-homeserver/src/core/database/tables/signup_tokens.rs:
```diff
@@ -0,0 +1,87 @@
+        Ok(token)
+    }
+
+    pub fn validate_signup_token(
+        &self,
+        token: &str,
+        user_pubkey: &PublicKey,
+    ) -> anyhow::Result<()> {
+        let mut wtxn = self.env.write_txn()?;
+        if let Some(token_bytes) = self.tables.signup_tokens.get(&wtxn, token)? {
+            let mut signup_token = SignupToken::deserialize(token_bytes);
+            if signup_token.used.is_some() {
+                anyhow::bail!("Token already used");
+            }
+            // Mark token as used.
+            signup_token.used = Some(user_pubkey.clone());
+            self.tables
+                .signup_tokens
+                .put(&mut wtxn, token, &signup_token.serialize())?;
+            wtxn.commit()?;
+            Ok(())
+        } else {
+            anyhow::bail!("Invalid token");
+        }
+    }
+}
```
Yes, makes sense!

SHAcollision commented on pubky-testnet/src/lib.rs:
```diff
@@ -107,6 +107,16 @@ impl Testnet {
+    /// Returns a [HomeserverBuilder] preconfigured with this testnet's DHT bootstrap nodes.
+    pub fn homeserver_builder(&self) -> HomeserverBuilder {
+        // Use the same test admin password as the run_test() homeserver.
+        let mut builder = Homeserver::builder().admin_password("test_admin_password".to_string());
```
sure!

SHAcollision commented on pubky/pkg/test/auth.js:
```diff
@@ -11,7 +11,7 @@ test('Auth: basic', async (t) => {
   const publicKey = keypair.publicKey()
 
-  await client.signup(keypair, HOMESERVER_PUBLICKEY )
+  await client.signup(keypair, HOMESERVER_PUBLICKEY, null)
```
Yes. It is not required here. I left here mostly as a pointer to come back and place an actual token. But first we need to implement a `pubky-testnet` with hardcoded configuration that ALSO requires a signup token so we can also add a test token signups from the nodejs/browser.

SHAcollision commented on pubky/src/native/api/auth.rs:
```diff
@@ -23,26 +23,51 @@ impl Client {
+        url.set_path("/signup");
+
+        // 2) If we have a signup_token, append it to the query string.
+        if let Some(token) = signup_token {
+            url.query_pairs_mut().append_pair("signup_token", token);
+        }
```
True, will fix.

SHAcollision replied:
Moving `signup_token` into the POST body makes sense to align with expected REST semantics (avoiding side effects in GET parameters). However, the current approach keeps AuthToken as raw bytes, and signup_token does not belong inside it. To handle this, we'd need to introduce a structured payload (e.g., JSON with `auth_token` as bytes and `signup_token` as an optional string), requiring additional dependencies (`serde`, `serde_bytes`) on both the homeserver and client.

Before committing to JSON, we should confirm whether this aligns with the original design choice of using raw bytes. Alternatively, we could send both as concatenated bytes, but that might be less clear. If we prefer to avoid unnecessary complexity for now, would it be acceptable to keep `signup_token` as a URL parameter until we fully commit to a structured request format?

SeverinAlexB replied:
I am fine with hacking it for the moment to deliver the feature but we should solve and refactor this soon.

SHAcollision commented:
I fixed the PR comments that were attainable by today. Those that are not fixed, I left a comment on the tradeoffs. We can take it slowly and in the meantime release `v0.4.2.rc1` 

I added instructions for generating signup tokens. These will now be needed on @catch-21 and @MiguelMedeiros e2e-tests . A default `pubky-testnet` homeserver can we queried to generate a new signup token like this:

from JS

```js
const url = `http://${homeserver_ipport}/admin/generate_signup_token`;
const response = await pubky_client.fetch(url, {   // would work with regular `fetch()` too
  method: "GET",
  headers: {
    "X-Admin-Password": "admin", // use your admin password, defaults to testnet password.
  },
});
const signupToken = await response.text();
```

via CLI with `curl`

```bash
curl -X GET "https://<homeserver_ip:port>/admin/generate_signup_token" \
     -H "X-Admin-Password: admin"
     # Use your admin password. This is testnet default pwd.
```
or from rust

```rust
let response = pubky_client
    .get(&format!("https://{homeserver_pubkey}/admin/generate_signup_token"))
    .header("X-Admin-Password", "admin") // Use your admin password. This is testnet default pwd.
    .send()
    .await
    .unwrap();
let signup_token = response.text().await.unwrap();
```

SeverinAlexB commented on pubky-homeserver/src/config.example.toml:
```diff
@@ -1,6 +1,14 @@
 # Secret key (in hex) to generate the Homeserver's Keypair
 # secret_key = "0000000000000000000000000000000000000000000000000000000000000000"
 
+[admin]
+# Set an admin password to protect admin endpoints.
+# If no password is set, the admin endpoints will not be accessible. 
+password = "admin"
+# Set signup_mode to "open" to allow anyone to signup a new user,
+# otherwise, "token_required" (the default) to require a valid invite token on signup;
+signup_mode = "token_required"
+
```
I don't like the `admin` section. What has the `signup_mode` to do with admin? Why not put both in the general section like the secret key?

We need to refactor the config.rs anyway in the future. So maybe we should merge it as it is and then do it better later?

SHAcollision replied:
Yes, I think there simply was not a right place to placed these two. Needs refactoring. So we could leave as is and refactor next.

SeverinAlexB commented on pubky-homeserver/src/core/database/tables/signup_tokens.rs:
```diff
@@ -0,0 +1,87 @@
+    pub token: String,
+    pub created_at: u64,
+    /// If Some(pubkey), the token has been used.
+    pub used: Option<PublicKey>,
```
NIT: rename to `used_by`? Not so important

SHAcollision replied:
We could rename, but that would be a breaking change against currently deployed prod homeserver that will need some sort of migration script. Maybe not worth it?

SeverinAlexB replied:
Let's keep it as it is and potentially address this when we look over the db setup again.

SeverinAlexB commented on pubky-homeserver/src/core/database/tables/signup_tokens.rs:
```diff
@@ -0,0 +1,87 @@
+        let mut wtxn = self.env.write_txn()?;
+        if let Some(token_bytes) = self.tables.signup_tokens.get(&wtxn, token)? {
+            let mut signup_token = SignupToken::deserialize(token_bytes);
+            if signup_token.used.is_some() {
+                anyhow::bail!("Token already used");
+            }
```
NIT: Create a method `is_used()` in `impl SignupToken`

SeverinAlexB commented on pubky-homeserver/src/core/routes/auth.rs:
```diff
@@ -1,79 +1,154 @@
-use crate::core::{database::tables::users::User, error::Result, AppState};
-
+/// Creates a brand-new user if they do not exist, then logs them in by creating a session.
+/// 1) Check if signup tokens are required (signup mode is Closed).
```
```suggestion
/// 1) Check if signup tokens are required (signup mode is token_required).
```

SeverinAlexB commented on pubky-homeserver/src/core/routes/auth.rs:
```diff
@@ -1,79 +1,154 @@
+    }
+    txn.commit()?;
+
+    // 3) If signup_mode == closed, require & validate a `signup_token` param.
```
```suggestion
    // 3) If signup_mode == token_required, require & validate a `signup_token` param.
```

SHAcollision replied:
Nice catches!

SeverinAlexB commented on pubky-homeserver/src/core/routes/auth.rs:
```diff
@@ -1,79 +1,154 @@
-                created_at: Timestamp::now().as_u64(),
-            },
-        )?;
+    let user_exists = users.get(&txn, public_key)?.is_some();
+    txn.commit()?;
+    if !user_exists {
+        return Err(Error::new(
+            StatusCode::NOT_FOUND,
+            Some("User does not exist"),
+        ));
     }
```
Comment, no need to take action: "does user exist" seems to be a common question and in the future, we should think of a extracting these methods in it's own struct next to the tables/database. Maybe a repository pattern?

SHAcollision replied:
Yes, agree

SeverinAlexB commented on pubky/pkg/test/utils.js:
```diff
@@ -0,0 +1,25 @@
+ * @returns {Promise<string>} - The signup token.
+ * @throws Will throw an error if the request fails.
+ */
+export async function getSignupToken(client, homeserver_address ="localhost:6286", adminPassword = "admin") {
```
`get` indicates that the token already exists. Can we do `createSignupToken`?

SHAcollision replied:
Yes, that makes sense

SeverinAlexB reviewed (changes requested):
We are getting there!
